app/recipe/tests/test_recipe_api.py

```diff
@@ -0,0 +1,177 @@
+from django.test import TestCase
+from django.urls import reverse
+
+from rest_framework import status
+
+from core.models import Recipe, Ingredient
+
+from recipe.serializers import RecipeDetailSerializer
+
+RECIPES_URL = reverse('recipe:recipe-list')
+
+
+def specific_recipe_url(recipe_id):
+    """Return recipe detail URL"""
+    return reverse('recipe:recipe-detail', args=[recipe_id])
+
+
+def sample_ingredient(recipe, name='Sample ingredient'):
+    """Create and return a sample ingredient"""
+    return Ingredient.objects.create(recipe=recipe, name=name)
+
+
+def sample_recipe(**params):
+    """Create and return a sample recipe"""
+    defaults = {
+        'name': 'Sample recipe',
+        'description': 'Sample recipe description'
+    }
+    defaults.update(params)
+
+    return Recipe.objects.create(**defaults)
+
+
+class RecipeApiTests(TestCase):
+
+    def test_retrieve_recipes(self):
+        """Test retrieving a list of recipes"""
+        recipe1 = sample_recipe(name='Steak pie')
+        sample_recipe(name='Pizza')
+
+        sample_ingredient(recipe1)
+
+        res = self.client.get(RECIPES_URL)
+
+        recipes = Recipe.objects.all().order_by('name')
+        serializer = RecipeDetailSerializer(recipes, many=True)
```

Comment:
I would assert the returned json in these tests and test the serializer separately in another test class.

For example for this test we could check we get 2 recipes:
```assert len(res.data) == 2```

And then have a separate test class where we check the serializer returns the json we expect for each field.

This is because serializers can get pretty complex, but it's not important for this "tk university" exercise

Reply:
oh nice, that makes sense, thank you. But then apart from having tests for the serializers, on the view tests should we also check that the returned json is fine apart from the length? Searching for examples I've seen that some people use the json dependency to transform the response.data (OrderedDict) to json, would that be the correct way of doing it?

Reply:
Yes, it does, thanks for the explanation. I agree, as we know that the view uses that specific serializer having the test for that one should be enough to ensure the json is fine! 
Is it fine if I leave the json assertion on this exercise as I haven't added the specific serializer tests? 
I'll have that in mind when working on our projects :)